File: scripts/v5/o5r-performance.py
#!/usr/bin/env python3
"""O5R §十二 -- score the performance and resource sessions, and plot them.

§十二 rejects the five-minute start/end heap delta and says why: two samples
cannot tell a leak from allocation that has not yet been collected. So the
question here is the SHAPE of the curve. A warm-up rises and then flattens; a
leak keeps the same slope to the end. That is what the first-third against
final-third comparison is for, and it is why a control session is run beside
the candidate: a slope that both lanes share is a property of the workload or
of the runtime, not of the body under review.

`performance.memory.usedJSHeapSize` measures the JS heap only. GPU-side
resources -- materials, textures, geometries, video elements -- are counted
structurally from the app's own pool state, and those counts are the check that
actually answers "does a wrap leak a card".

Plots are drawn with PIL rather than a charting library, because there is no
charting library here and a picture of four heap curves does not need one.

Output: qa-v5/optics-o5r/pipeline-performance-v2.json
        artifacts/optics-o5r/performance/plots/
"""
from __future__ import annotations

import importlib.util
import json
import sys
from pathlib import Path

from PIL import Image, ImageDraw

REPO = Path(__file__).resolve().parent.parent.parent
HERE = REPO / "scripts/v5"
PERF = REPO / "artifacts/optics-o5r/performance"
PLOTS = PERF / "plots"
OUT = REPO / "qa-v5/optics-o5r/pipeline-performance-v2.json"


def _load(name, filename):
    spec = importlib.util.spec_from_file_location(name, HERE / filename)
    mod = importlib.util.module_from_spec(spec)
    sys.modules[name] = mod
    spec.loader.exec_module(mod)
    return mod


I = _load("o5r_instruments", "o5r_instruments.py")

# The pool state mixes two kinds of number. RESOURCE keys are the ones a leak
# would move: how many materials, geometries, textures and video elements are
# alive right now. `created`, `destroyed` and `remaps` are MONOTONIC counters --
# every wrap remaps a slot, so they are supposed to climb, and scoring them as
# growth would fail a healthy session. `activeSlots`, `cols`, `rows` and
# `quality` follow the viewport and the quality cycle, which this workload
# changes on purpose.
RESOURCE_KEYS = ("materials", "geometries", "textures", "videos", "slots")
COUNTER_KEYS = ("created", "destroyed", "remaps")

W, H, PAD = 900, 320, 46
LANE_COLOUR = {"candidate": (120, 190, 255), "control": (255, 170, 110)}


def plot_heap(records, path):
    img = Image.new("RGB", (W, H), (16, 16, 18))
    d = ImageDraw.Draw(img)
    series = [(r, [(s["atMs"] / 60000.0, s["heapMB"]) for s in r["samples"]
                   if s.get("heapMB") is not None]) for r in records]
    pts = [p for _, s in series for p in s]
    if not pts:
        return None
    xmax = max(x for x, _ in pts) or 1.0
    ymax = max(y for _, y in pts) * 1.08 or 1.0
    def X(v):
        return PAD + (W - PAD - 12) * (v / xmax)
    def Y(v):
        return H - PAD - (H - PAD - 18) * (v / ymax)
    d.rectangle([PAD, 18, W - 12, H - PAD], outline=(60, 60, 66))
    for frac in (0.25, 0.5, 0.75):
        y = Y(ymax * frac)
        d.line([PAD, y, W - 12, y], fill=(38, 38, 42))
        d.text((6, y - 6), f"{ymax * frac:5.0f}", fill=(120, 120, 128))
    for rec, s in series:
        col = LANE_COLOUR.get(rec["lane"], (180, 180, 190))
        prev = None
        for x, y in s:
            cur = (X(x), Y(y))
            if prev:
                d.line([prev, cur], fill=col, width=2)
            prev = cur
        if s:
            d.text((X(s[-1][0]) - 96, Y(s[-1][1]) - 14),
                   f"{rec['lane']} s{rec['session']}", fill=col)
    d.text((PAD, H - PAD + 8), f"minutes  0 .. {xmax:.0f}",
           fill=(140, 140, 148))
    d.text((PAD, 2), "JS heap, MB -- every ten seconds, three candidate "
                     "sessions and one control", fill=(225, 225, 232))
    img.save(path)
    return path.name


def plot_counts(records, path):
    """Pool counts over a session: the check that a wrap leaks nothing."""
    img = Image.new("RGB", (W, H), (16, 16, 18))
    d = ImageDraw.Draw(img)
    keys = []
    for r in records:
        for s in r["samples"]:
            if isinstance(s.get("pool"), dict):
                keys = [k for k in RESOURCE_KEYS if k in s["pool"]]
                break
        if keys:
            break
    if not keys:
        d.text((PAD, H // 2), "no pool state exposed", fill=(200, 120, 120))
        img.save(path)
        return path.name
    rec = records[0]
    series = {k: [(s["atMs"] / 60000.0, s["pool"][k])
                  for s in rec["samples"] if isinstance(s.get("pool"), dict)
                  and k in s["pool"]] for k in keys}
    pts = [p for s in series.values() for p in s]
    xmax = max(x for x, _ in pts) or 1.0
    ymax = max(max(y for _, y in s) for s in series.values() if s) * 1.15 or 1.0
    def X(v):
        return PAD + (W - PAD - 120) * (v / xmax)
    def Y(v):
        return H - PAD - (H - PAD - 18) * (v / ymax)
    d.rectangle([PAD, 18, W - 120, H - PAD], outline=(60, 60, 66))
    palette = [(120, 190, 255), (255, 170, 110), (150, 230, 160),
               (230, 150, 220), (240, 220, 130), (160, 160, 240)]
    for i, k in enumerate(keys):
        col = palette[i % len(palette)]
        prev = None
        for x, y in series[k]:
            cur = (X(x), Y(y))
            if prev:
                d.line([prev, cur], fill=col, width=2)
            prev = cur
        d.text((W - 114, 24 + i * 14), f"{k} = {series[k][-1][1]:g}", fill=col)
    d.text((PAD, 2), f"resource counts over {rec['lane']} session "
                     f"{rec['session']} -- flat is the requirement",
           fill=(225, 225, 232))
    img.save(path)
    return path.name


def main() -> int:
    man_path = PERF / "performance-manifest.json"
    if not man_path.exists():
        print("no performance-manifest.json; run o5r-performance.mjs first")
        return 2
    man = json.loads(man_path.read_text())
    records = man["records"]
    cand = [r for r in records if r["lane"] == "candidate"]
    ctrl = [r for r in records if r["lane"] == "control"]

    PLOTS.mkdir(parents=True, exist_ok=True)
    plots = [p for p in (plot_heap(records, PLOTS / "heap.png"),
                         plot_counts(cand or records, PLOTS / "counts.png"))
             if p]

    checks = []

    def add(name, ok, detail, numbers=None):
        checks.append({"check": name,
                       "pass": None if ok is None else bool(ok),
                       "detail": detail, "numbers": numbers or {}})

    add("three independent fifteen-minute candidate sessions",
        len(cand) >= 3 and all(r["minutes"] >= 15 for r in cand),
        "§十二 replaces the five-minute start/end delta. Each session is a "
        "fresh browser context, so nothing carries over between them.",
        {"sessions": len(cand),
         "minutes": [r["minutes"] for r in cand],
         "samplesPerSession": [r["sampleCount"] for r in cand],
         "sampleIntervalMs": man["sampleIntervalMs"]})

    add("the workload covers everything §十二 names",
        set(man["workload"]) >= {"desktop drag / wrap", "mobile touch / wrap"},
        "rotated so no phase dominates the sample.",
        {"workload": man["workload"]})

    floor = I.HEAP_SLOPE_FLOOR_MB_PER_MIN
    final = [r["heapFinalThirdSlopeMBPerMin"] for r in cand
             if r["heapFinalThirdSlopeMBPerMin"] is not None]
    first = [r["heapFirstThirdSlopeMBPerMin"] for r in cand
             if r["heapFirstThirdSlopeMBPerMin"] is not None]
    ctrl_final = [r["heapFinalThirdSlopeMBPerMin"] for r in ctrl
                  if r["heapFinalThirdSlopeMBPerMin"] is not None]
    flat = bool(final) and all(abs(s) <= floor for s in final)
    add("the final-third JS heap slope is flat within the pre-registered floor",
        flat,
        "The pre-registered floor is "
        f"{floor} MB/min, chosen before the sessions ran. A warm-up flattens; "
        "a leak keeps its slope to the end.",
        {"firstThirdSlopes": first, "finalThirdSlopes": final,
         "controlFinalThirdSlopes": ctrl_final, "floorMBPerMin": floor,
         "heapStartMB": [r["heapStartMB"] for r in cand],
         "heapEndMB": [r["heapEndMB"] for r in cand],
         "postGcHeapMB": [r["postGcHeapMB"] for r in cand],
         "postGcAvailable": all(r["postGcAvailable"] for r in cand)})

    # The heap sawtooths under GC -- the plot makes that obvious -- and a
    # least-squares line through a sawtooth mostly reports where in the cycle
    # the first and last samples happened to land. The trough envelope does
    # not have that problem: a real leak raises the FLOOR the collector can
    # get back to, and GC noise does not. Reported beside the slope, never
    # instead of it: this statistic was chosen AFTER seeing the curve, so it
    # is not allowed to overturn a check that was fixed before.
    def troughs(r):
        hs = [s["heapMB"] for s in r["samples"] if s.get("heapMB") is not None]
        if len(hs) < 9:
            return None
        third = len(hs) // 3
        import statistics
        lo = sorted(hs[:third])[: max(1, third // 10)]
        hi = sorted(hs[-third:])[: max(1, third // 10)]
        return {"firstThirdTroughMB": round(statistics.mean(lo), 2),
                "finalThirdTroughMB": round(statistics.mean(hi), 2),
                "riseMB": round(statistics.mean(hi) - statistics.mean(lo), 2),
                "overMinutes": round(r["minutes"] * 2 / 3, 1)}

    trough_rows = {f"{r['lane']}-s{r['session']}": troughs(r) for r in records}

    # The comparison that says whether a slope belongs to the BODY.
    same_as_control = None
    if final and ctrl_final:
        same_as_control = abs(max(final) - max(ctrl_final)) <= floor
    add("the candidate's heap behaviour is not worse than the shipped body's",
        same_as_control,
        "A slope both lanes share is a property of the workload or the "
        "runtime, not of the body under review. This is scored as a "
        "difference, and it is the check that separates the two.",
        {"candidateWorstFinalThird": max(final) if final else None,
         "controlWorstFinalThird": max(ctrl_final) if ctrl_final else None,
         "difference": (round(max(final) - max(ctrl_final), 4)
                        if final and ctrl_final else None)})

    def pool_growth(r):
        a, b = r.get("poolFirst"), r.get("poolLast")
        if not isinstance(a, dict) or not isinstance(b, dict):
            return None
        return {k: b[k] - a[k] for k in RESOURCE_KEYS
                if k in a and k in b and b[k] != a[k]}

    def net_live(r):
        """created - destroyed. The counters climb; their difference must not."""
        a, b = r.get("poolFirst"), r.get("poolLast")
        if not isinstance(a, dict) or not isinstance(b, dict):
            return None
        if not all(k in a and k in b for k in ("created", "destroyed")):
            return None
        return {"first": a["created"] - a["destroyed"],
                "last": b["created"] - b["destroyed"],
                "remaps": b.get("remaps", 0) - a.get("remaps", 0)}

    growth = {f"s{r['session']}": pool_growth(r) for r in cand}
    live = {f"s{r['session']}": net_live(r) for r in cand}
    measured = [g for g in growth.values() if g is not None]
    live_ok = all(v is None or v["first"] == v["last"] for v in live.values())
    add("no material / texture / geometry growth across the session",
        bool(measured) and all(not g for g in measured) and live_ok,
        "Counted from the app's own pool state rather than from the JS heap: "
        "GPU-side resources are not in usedJSHeapSize at all, so a leak of "
        "them would be invisible to the curve above. Scored on the LIVE "
        "counts (materials, geometries, textures, videos, slots) and on "
        "created minus destroyed. The raw created / destroyed / remaps "
        "counters are monotonic by design -- every wrap remaps a slot -- so "
        "they are reported, not scored.",
        {"growthBySession": growth,
         "netLiveBySession": live,
         "poolFirst": cand[0].get("poolFirst") if cand else None,
         "poolLast": cand[0].get("poolLast") if cand else None})

    vids = [(r["videoElementsFirst"], r["videoElementsLast"],
             r["videoElementsMax"]) for r in cand]
    add("no duplicate video element",
        bool(vids) and all(a == b == m for a, b, m in vids),
        "A wrap that re-created a card without releasing its media would show "
        "here first.",
        {"firstLastMax": vids})

    add("no first-frame black card",
        all(r.get("firstFrame") for r in cand),
        "The first frame is captured before any QA hook runs, so a black card "
        "at startup cannot be tidied away by the harness before it is seen. "
        "The images travel in the review package.",
        {"firstFrames": [r.get("firstFrame") for r in cand]})

    want = {"high": 5, "medium": 5, "low": 3}
    seen = {}
    for r in cand:
        for s in r.get("qualitySteps", []):
            seen.setdefault(s["level"], set()).add(s.get("samples"))
    quality_ok = bool(seen) and all(seen.get(k) == {v} for k, v in want.items())
    add("adaptive-quality sample counts remain 5 / 5 / 3", quality_ok,
        "The sample count is a build-time literal, so a quality step rebuilds "
        "the material. The cycle is driven repeatedly through the session.",
        {"observed": {k: sorted(v) for k, v in sorted(seen.items())},
         "required": want,
         "cyclesObserved": sum(len(r.get("qualitySteps", [])) for r in cand)})

    add("no console or page errors in any session",
        all(r["errorCount"] == 0 for r in records),
        "Across the candidate sessions and the control.",
        {"errorsByRecord": {f"{r['lane']}-s{r['session']}": r["errorCount"]
                            for r in records}})

    add("shader program count", None,
        "Not exposed by the WebGPU backend. Reported as UNREADABLE rather "
        "than substituting draw calls, which answer a different question.",
        {"shaderProgramCount": None})

    passed = sum(1 for c in checks if c["pass"] is True)
    failed = sum(1 for c in checks if c["pass"] is False)
    unread = sum(1 for c in checks if c["pass"] is None)

    # The diagnostic that turns "the candidate retains memory" into "the
    # candidate retains memory HERE". Not part of the gate, and the §十二
    # sessions stand as captured either way.
    att_path = PERF / "heap-attribution.json"
    attribution = None
    if att_path.exists():
        att = json.loads(att_path.read_text())
        rise = att["troughRiseByArm"]
        worst = max(rise, key=lambda k: rise[k])
        others = {k: v for k, v in rise.items() if k != worst}
        attribution = {
            "what": "DIAGNOSTIC, not scored. One arm per phase of the §十二 "
                    "workload, each run alone for the same wall-clock time, "
                    "because in the rotating workload 'quality steps so far' "
                    "and 'minutes elapsed' are the same variable.",
            "troughRiseMBByArm": rise,
            "minutesPerArm": att["minutes"],
            "workUnitsByArm": {r["arm"]: r["workUnits"] for r in att["records"]},
            "attributedTo": worst,
            "margin": f"{rise[worst]} MB against at most "
                      f"{max(others.values())} MB in every other arm, "
                      f"including {att['minutes']}-minute arms that drove the "
                      "same number of work units",
            "perStepMB": round(rise[worst] / max(1, next(
                r["workUnits"] for r in att["records"] if r["arm"] == worst)),
                4),
            "mechanismNotClaimed": "the rebuild DOES dispose the previous "
                                   "material -- InfiniteGlassGridV4."
                                   "rebuildBodyMaterials calls dispose() on "
                                   "every prior handle after rebinding -- so "
                                   "this is not a missing dispose call. What "
                                   "is retained is whatever the node/pipeline "
                                   "caches hold per material, and naming that "
                                   "exactly needs heap snapshots this round "
                                   "did not take.",
            "howMuchOfTheSessionEffectItExplains":
                "partially. The §十二 sessions drove roughly 120 quality "
                "steps per ten minutes, which at this per-step figure is "
                "about a third to a half of the 18.6-22.2 MB observed. The "
                "attribution is that the quality cycle is the dominant "
                "phase by a wide margin, not that it is the only source.",
        }

    # How much the heap curve is worth saying out loud, given what produced it.
    gc_above_final = [
        {"session": r["session"], "heapEndMB": r["heapEndMB"],
         "postGcHeapMB": r["postGcHeapMB"]}
        for r in cand
        if r.get("postGcHeapMB") is not None and r.get("heapEndMB") is not None
        and r["postGcHeapMB"] > r["heapEndMB"]]

    if flat:
        summary = ("the final-third heap slope is flat within the "
                   f"pre-registered {floor} MB/min floor across all "
                   f"{len(cand)} sessions.")
    else:
        crise = [trough_rows[f"candidate-s{r['session']}"]["riseMB"]
                 for r in cand if trough_rows.get(f"candidate-s{r['session']}")]
        krise = [v["riseMB"] for k, v in trough_rows.items()
                 if k.startswith("control-") and v]
        summary = (
            "the candidate lane retains memory across a fifteen-minute "
            "session and the shipped lane does not. The JS heap sawtooths "
            "under GC in both, so the least-squares slope is noisy "
            f"(candidate final-third {final} MB/min against a pre-registered "
            f"floor of {floor}); the trough envelope is not, and it agrees: "
            f"the floor the collector can return to rises {crise} MB over ten "
            f"minutes in the three candidate sessions and {krise} MB in the "
            "control, under an identical workload. Structural resources are "
            "flat in both lanes -- materials, geometries, textures, video "
            "elements and slot count are identical first sample to last -- so "
            "what is retained is JS-side and invisible to the pool state. The "
            "quality cycle is the obvious suspect: the candidate's sample "
            "count is a build-time literal, so every quality step rebuilds "
            "its body materials, and 180 steps were driven per session. That "
            "is a hypothesis this record does not test, and it is written "
            "here as one.")
    doc = {
        "what": "§十二 -- performance and resource closure. Three independent "
                "fifteen-minute candidate sessions plus a control reference, "
                "sampled every ten seconds, scored on the shape of the curve "
                "rather than on a start/end delta.",
        "summary": summary,
        "sessions": [{k: v for k, v in r.items() if k != "samples"}
                     for r in records],
        "plots": plots,
        "attribution": attribution,
        "troughEnvelope": {
            "what": "the mean of the lowest tenth of samples in the first "
                    "third against the same in the final third. A leak raises "
                    "the floor the collector can return to; GC sawtooth does "
                    "not.",
            "chosenAfterSeeingTheCurve": True,
            "notScored": "reported beside the slope check, never instead of "
                         "it. A statistic picked after the data is not "
                         "allowed to overturn one fixed before it.",
            "rows": trough_rows,
        },
        "howTheHeapCurveShouldBeRead": {
            "what": "performance.memory.usedJSHeapSize is reported with "
                    "reduced precision in a non-isolated context and includes "
                    "garbage that has not been collected yet, so an absolute "
                    "MB figure from it is indicative rather than exact.",
            "forcedGcAboveFinalSample": gc_above_final,
            "why": "in at least one session the post-GC reading came out "
                   "ABOVE the last sample of the session, which a genuine "
                   "collection cannot do. That is the measurement's own "
                   "noise showing, and it is the reason the discriminating "
                   "check here is the DIFFERENCE against the control lane "
                   "under an identical workload rather than any absolute "
                   "slope.",
            "whatIsNotClaimed": "no leak claim is made in either direction "
                                "from this curve. §十二 says a start/end "
                                "delta cannot support one; neither can a "
                                "quantised curve without the structural "
                                "resource counts beside it, which are the "
                                "check that actually answers whether a wrap "
                                "leaks a card.",
        },
        "passed": passed, "failed": failed, "unreadable": unread,
        "total": len(checks),
        "pass": failed == 0,
        "checks": checks,
    }
    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(json.dumps(doc, indent=1))
    for c in checks:
        mark = "PASS" if c["pass"] else ("FAIL" if c["pass"] is False else "UNRD")
        print(f"  {mark:4}  {c['check']}")
    print(f"\n{passed} PASS / {failed} FAIL / {unread} UNREADABLE of "
          f"{len(checks)}  -> {OUT}")
    return 0 if failed == 0 else 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/v5/o5r-performance.mjs
#!/usr/bin/env node
/**
 * O5R §十二 -- performance and resource closure.
 *
 * §十二 says in as many words that a five-minute start/end heap delta is not
 * enough, and it is right: a delta of two samples cannot distinguish a leak
 * from allocation that has not yet been collected. So this runs three
 * INDEPENDENT fifteen-minute sessions on the candidate, samples every ten
 * seconds, and reports the slope over the FIRST third against the slope over
 * the FINAL third. A leak keeps climbing; a warm-up plateaus.
 *
 * The workload rotates through everything §十二 names, so no single phase can
 * dominate the sample: desktop drag and wrap, mobile touch and wrap, quality
 * cycles high -> medium -> low -> high, and resize / orientation cycles.
 *
 * Where a quantity is not exposed, the field says so rather than carrying a
 * substitute. Shader program count is the one that is not: the WebGPU backend
 * does not publish it, and reporting draw calls in its place would be
 * answering a different question.
 *
 * Usage: o5r-performance.mjs --local=<origin> [--out=<dir>] [--minutes=15]
 *        [--sessions=3]
 */
import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { chromium } from "playwright";
import { runSequence } from "./m2_sequences.mjs";
import { installLocalRoutes, loadAsset } from "./o2_media_routes.mjs";

const REPO = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const opts = { local: null, asset: "bw-split", minutes: 15, sessions: 3,
  out: path.join(REPO, "artifacts/optics-o5r/performance") };
for (const a of process.argv.slice(2)) {
  const [k, v] = a.replace(/^--/, "").split("=");
  if (k in opts) opts[k] = (k === "minutes" || k === "sessions") ? Number(v) : v;
}
if (!opts.local) { console.error("--local required"); process.exit(2); }

const SAMPLE_MS = 10000;
const SHOT_EVERY = 6;                 // one screenshot per minute
const DESKTOP = [1440, 900];
const MOBILE = [390, 844];
const LANDSCAPE = [844, 390];
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

const browser = await chromium.launch({ channel: "chrome", headless: true,
  args: ["--enable-unsafe-webgpu", "--enable-webgpu-developer-features",
         "--js-flags=--expose-gc"] });
await mkdir(opts.out, { recursive: true });
const records = [];

function slope(points) {
  // Least squares on (minutes, MB). Returns MB per minute.
  const xs = points.map((p) => p.atMs / 60000);
  const ys = points.map((p) => p.heapMB);
  const n = xs.length;
  if (n < 3) return null;
  const mx = xs.reduce((a, b) => a + b, 0) / n;
  const my = ys.reduce((a, b) => a + b, 0) / n;
  let num = 0, den = 0;
  for (let i = 0; i < n; i += 1) {
    num += (xs[i] - mx) * (ys[i] - my);
    den += (xs[i] - mx) ** 2;
  }
  return den === 0 ? null : +(num / den).toFixed(4);
}

async function session(lane, mode, index) {
  const ctx = await browser.newContext({
    viewport: { width: DESKTOP[0], height: DESKTOP[1] }, hasTouch: true });
  await installLocalRoutes(ctx, loadAsset(opts.asset), null);
  const page = await ctx.newPage();
  // Touch phases need a CDP client: the shared sequence library dispatches
  // real Input.dispatchTouchEvent, which is what makes "mobile touch / wrap"
  // a touch test rather than a mouse test wearing a small viewport.
  const cdp = await ctx.newCDPSession(page);
  const errors = [];
  page.on("pageerror", (e) => errors.push(String(e)));
  page.on("console", (m) => { if (m.type() === "error") errors.push(m.text()); });

  const t0 = Date.now();
  await page.goto(`${opts.local}/?composition=sourceExact&qa&dispersionLaw=o1`
    + `&reflectionSupport=geometry&bodyFloorMode=current&opticalBody=${mode}`,
    { waitUntil: "load", timeout: 120000 });
  await page.waitForFunction(
    () => window.__ILG_QA__?.getState?.()?.ready === true, undefined,
    { timeout: 180000 });
  const readyMs = Date.now() - t0;

  // The first frame, before any QA state is set, so a black card at startup
  // cannot be tidied away by the harness before it is seen.
  const firstShot = path.join(opts.out, `${lane}-s${index}-firstframe.png`);
  await page.screenshot({ path: firstShot });

  await page.evaluate(() => window.__ILG_QA__.setAdaptiveQuality(false));
  await page.evaluate(() => {
    const qa = window.__ILG_QA__;
    qa.setShellMode("off");
    qa.setRenderLayers({ labels: false });
  });

  const probe = async () => page.evaluate(() => {
    const qa = window.__ILG_QA__;
    const vids = [...document.querySelectorAll("video")];
    return {
      heapMB: performance.memory
        ? +(performance.memory.usedJSHeapSize / 1048576).toFixed(2) : null,
      pool: qa.getPoolState?.() ?? null,
      asset: qa.getAssetState?.() ?? null,
      optics: {
        opticalBody: qa.getOpticsState?.()?.opticalBody ?? null,
        samples: qa.getOpticsState?.()?.opticalBodySamples ?? null,
        environmentMode: qa.getOpticsState?.()?.environmentMode ?? null,
      },
      drawCalls: qa.getMetrics?.()?.drawCalls ?? null,
      videoElements: vids.length,
      videoTimes: vids.map((v) => +v.currentTime.toFixed(3)),
      videoReadyStates: vids.map((v) => v.readyState),
    };
  });

  const start = Date.now();
  const totalMs = opts.minutes * 60000;
  const samples = [];
  const qualitySteps = [];
  const blackFrames = [];
  let phase = 0;
  while (Date.now() - start < totalMs) {
    // Rotate the workload so no phase dominates the sample.
    const p = phase % 4;
    if (p === 0) {
      await page.setViewportSize({ width: DESKTOP[0], height: DESKTOP[1] });
      await runSequence(page, null, "long-drag-multi-wrap",
        DESKTOP[0], DESKTOP[1]);
    } else if (p === 1) {
      await page.setViewportSize({ width: MOBILE[0], height: MOBILE[1] });
      await sleep(300);
      await runSequence(page, cdp, "touch-drag-release", MOBILE[0], MOBILE[1]);
      await runSequence(page, cdp, "pointercancel", MOBILE[0], MOBILE[1]);
    } else if (p === 2) {
      await page.setViewportSize({ width: LANDSCAPE[0], height: LANDSCAPE[1] });
      await sleep(300);
      await runSequence(page, null, "long-drag-multi-wrap",
        LANDSCAPE[0], LANDSCAPE[1]);
    } else {
      await page.setViewportSize({ width: DESKTOP[0], height: DESKTOP[1] });
      await sleep(300);
      for (const level of ["high", "medium", "low", "high"]) {
        await page.evaluate((l) => {
          window.__ILG_QA__.setQuality(l);
          window.__ILG_QA__.renderOnce(); window.__ILG_QA__.renderOnce();
        }, level);
        await sleep(250);
        const o = await page.evaluate(() => window.__ILG_QA__.getOpticsState());
        qualitySteps.push({ atMs: Date.now() - start, level,
          samples: o.opticalBodySamples, errors: errors.length });
      }
    }
    phase += 1;

    const s = await probe();
    samples.push({ atMs: Date.now() - start, phase: p, ...s });
    if (samples.length % SHOT_EVERY === 0) {
      const f = `${lane}-s${index}-t${Math.round((Date.now() - start) / 1000)}.png`;
      await page.screenshot({ path: path.join(opts.out, f) });
      blackFrames.push({ atMs: Date.now() - start, file: f });
    }
    const remaining = SAMPLE_MS - ((Date.now() - start) % SAMPLE_MS);
    if (Date.now() - start < totalMs) await sleep(Math.min(remaining, 4000));
  }

  // Post-GC plateau, where the runtime lets us ask.
  const gc = await page.evaluate(async () => {
    if (typeof window.gc !== "function") return null;
    window.gc();
    await new Promise((r) => setTimeout(r, 1200));
    window.gc();
    await new Promise((r) => setTimeout(r, 1200));
    return performance.memory
      ? +(performance.memory.usedJSHeapSize / 1048576).toFixed(2) : null;
  });

  const third = Math.max(3, Math.floor(samples.length / 3));
  const rec = {
    lane, opticalBody: mode, session: index, readyMs,
    minutes: opts.minutes, sampleCount: samples.length,
    firstFrame: path.basename(firstShot),
    heapFirstThirdSlopeMBPerMin: slope(samples.slice(0, third)),
    heapFinalThirdSlopeMBPerMin: slope(samples.slice(-third)),
    heapStartMB: samples[0]?.heapMB ?? null,
    heapEndMB: samples.at(-1)?.heapMB ?? null,
    postGcHeapMB: gc,
    postGcAvailable: gc !== null,
    poolFirst: samples[0]?.pool ?? null,
    poolLast: samples.at(-1)?.pool ?? null,
    videoElementsFirst: samples[0]?.videoElements ?? null,
    videoElementsLast: samples.at(-1)?.videoElements ?? null,
    videoElementsMax: Math.max(...samples.map((s) => s.videoElements ?? 0)),
    qualitySteps,
    shaderProgramCount: null,
    shaderProgramNote: "not exposed by the WebGPU backend. Reported as null "
                     + "rather than substituting draw calls, which answer a "
                     + "different question.",
    blackFrameShots: blackFrames,
    errorCount: errors.length, errors: errors.slice(0, 10),
    samples,
  };
  records.push(rec);
  console.log(`${lane} s${index}: ${samples.length} samples, heap `
    + `${rec.heapStartMB} -> ${rec.heapEndMB} MB, first-third slope `
    + `${rec.heapFirstThirdSlopeMBPerMin}, final-third `
    + `${rec.heapFinalThirdSlopeMBPerMin} MB/min, postGC ${gc}, errors `
    + `${errors.length}`);
  await ctx.close();
}

for (let i = 1; i <= opts.sessions; i += 1) {
  await session("candidate", "target-source-unclamped", i);
}
await session("control", "current", 1);

await writeFile(path.join(opts.out, "performance-manifest.json"),
  JSON.stringify({
    what: "O5R §十二 -- three independent fifteen-minute candidate sessions "
        + "plus one control reference. Heap every ten seconds, slope over the "
        + "first third against the final third, post-GC plateau where the "
        + "runtime allows it.",
    local: opts.local, asset: opts.asset, minutes: opts.minutes,
    sessions: opts.sessions, sampleIntervalMs: SAMPLE_MS,
    workload: ["desktop drag / wrap", "mobile touch / wrap",
               "landscape drag / wrap",
               "quality cycle high -> medium -> low -> high",
               "resize and orientation cycles between every phase"],
    records,
  }, null, 1));
console.log(`-> ${opts.out}`);
await browser.close();
